src/serializers/encoder/v7/encode_graph.c

```diff
@@ -43,28 +42,96 @@ static void _RdbSaveHeader(RedisModuleIO *rdb, GraphContext *gc) {
+		break;
+	case EDGES:
+		required_entities_count = Graph_EdgeCount(gc->g);
+		break;
+	case DELETED_EDGES:
+		required_entities_count = Graph_DeletedEdgeCount(gc->g);
+		break;
+	case GRAPH_SCHEMA:
+		required_entities_count = 0;
+		break;
+	default:
+		assert(false && "Unkown encoding state in _CurrentStatePayloadInfo");
+	}
+	PayloadInfo payload_info;
+	payload_info.state = state;
+	/* When this state will be encoded, the number of entities to encode is the minimum between the number of entities to encode and
+	 * the remaining entities left to encode from the same type. */
+	payload_info.entities_count = MIN(entities_to_encode, required_entities_count - offset);
+	return payload_info;
+}
+
+// This function save the key content schema and returns it so the encoder can now how to encode the key.
+static PayloadInfo *_RdbSaveKeySchema(RedisModuleIO *rdb, GraphContext *gc) {
+	/* Format:
+	*  #Number of payloads info - N
+	*  N * Payload info:
+	*      Encode state
+	*      Number of entities encoded in this state.
+	*/
+	PayloadInfo *payloads = array_new(PayloadInfo, 1);
+	// Get current encoding state.
+	EncodeState current_state = GraphEncodeContext_GetEncodeState(gc->encoding_context);
+	// If it is the start of the encodeing, set the state to be NODES.
+	if(current_state == INIT) current_state = NODES;
+	uint64_t remaining_entities = Config_GetModuleConfig().entities_threshold;
+	// No limit on the entities, the graph is encoded in one key.
+	if(remaining_entities == UNLIMITED) {
+		PayloadInfo nodes_info = {.state = NODES, .entities_count = Graph_NodeCount(gc->g)};
+		PayloadInfo deleted_nodes_info = {.state = DELETED_NODES, .entities_count = Graph_DeletedNodeCount(gc->g)};
+		PayloadInfo edges_info = {.state = EDGES, .entities_count = Graph_EdgeCount(gc->g)};
+		PayloadInfo deleted_edges_info = {.state = DELETED_EDGES, .entities_count = Graph_DeletedEdgeCount(gc->g)};
+		PayloadInfo graph_schema_info = {.state = GRAPH_SCHEMA, .entities_count = 0};
+		payloads = array_append(payloads, nodes_info);
+		payloads = array_append(payloads, deleted_nodes_info);
+		payloads = array_append(payloads, edges_info);
+		payloads = array_append(payloads, deleted_edges_info);
+		payloads = array_append(payloads, graph_schema_info);
+	} else {
+		// Get the current state encoded entities count.
+		uint64_t offset = GraphEncodeContext_GetProcessedEntitiesCount(gc->encoding_context);
+		// While there are still remaining entities to encode in this key and the state is valid.
+		while(remaining_entities > 0 && current_state <= GRAPH_SCHEMA) {
+			// Get the current state payload info, with respect to offset.
+			PayloadInfo current_state_payload_info = _StatePayloadInfo(gc, current_state, offset,
+																	   remaining_entities);
+			payloads = array_append(payloads, current_state_payload_info);
+			remaining_entities -= current_state_payload_info.entities_count;
+			current_state++; // Advance in the states.
+			offset = 0; // New state offset is 0.
```

Comment:
I understand why it is OK to update both `current_state` and `offset` but if the case where `remaining_entities = 0` as a result of too many entities in the current state, performing the update is unnecessary and confusing.